File: backend/domain/agentthread/entity/journal.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import "strings"

const (
	JournalSchemaVersion  = "1.1"
	JournalPayloadVersion = "1.0"
)

type RunAttemptStatus string

const (
	RunAttemptStatusPending     RunAttemptStatus = "pending"
	RunAttemptStatusRunning     RunAttemptStatus = "running"
	RunAttemptStatusCompleted   RunAttemptStatus = "completed"
	RunAttemptStatusFailed      RunAttemptStatus = "failed"
	RunAttemptStatusCancelled   RunAttemptStatus = "cancelled"
	RunAttemptStatusTimedOut    RunAttemptStatus = "timed_out"
	RunAttemptStatusInterrupted RunAttemptStatus = "interrupted"

	// Compatibility aliases keep existing internal callers source-compatible
	// while persisting only the frozen wire values above.
	RunAttemptStatusActive   = RunAttemptStatusRunning
	RunAttemptStatusCanceled = RunAttemptStatusCancelled
)

func (s RunAttemptStatus) IsActive() bool {
	return s == RunAttemptStatusPending || s == RunAttemptStatusRunning
}

func (s RunAttemptStatus) IsTerminal() bool {
	switch s {
	case RunAttemptStatusCompleted, RunAttemptStatusFailed,
		RunAttemptStatusCancelled, RunAttemptStatusTimedOut:
		return true
	default:
		return false
	}
}

// IsJournalTimeoutErrorCode recognizes only server-owned timeout categories.
func IsJournalTimeoutErrorCode(code string) bool {
	switch strings.ToLower(strings.TrimSpace(code)) {
	case "deadline_exceeded", "run_timeout", "task_timeout", "tool_timeout", "subagent_timeout", "timeout":
		return true
	default:
		return false
	}
}

type JournalProjectionState string

const (
	JournalProjectionStateHealthy  JournalProjectionState = "healthy"
	JournalProjectionStateDegraded JournalProjectionState = "degraded"
	JournalProjectionStateDisabled JournalProjectionState = "disabled"
)

type JournalVisibility string

const (
	JournalVisibilityUser     JournalVisibility = "user"
	JournalVisibilityInternal JournalVisibility = "internal"

	// Public was the pre-freeze internal name. It now aliases the user scope.
	JournalVisibilityPublic = JournalVisibilityUser
)

type RunAttempt struct {
	ID                     int64
	ThreadID               int64
	JournalRunID           int64
	ExecutionRunID         int64
	AttemptID              string
	Ordinal                uint32
	Status                 RunAttemptStatus
	ActiveSlot             *uint8
	NextSequence           uint64
	LastCommittedSequence  uint64
	SourceCheckpointID     *int64
	SourceAttemptID        *string
	RecoveryIdempotencyKey *string
	EnrollmentVersion      string
	SnapshotsEnabled       bool
	ProjectionState        JournalProjectionState
	ProjectionDegradedAt   *int64
	TraceID                *string
	TerminalEventID        *int64
	CreatedAt              int64
	UpdatedAt              int64
	StartedAt              *int64
	EndedAt                *int64
}

type JournalEvent struct {
	ID                 int64
	ThreadID           int64
	RunID              int64
	JournalRunID       int64
	AttemptID          string
	Sequence           uint64
	IdempotencyKey     string
	ParentEventID      int64
	SchemaVersion      string
	Status             string
	OccurredAtUnixNano int64
	Visibility         JournalVisibility
	PayloadVersion     string
	SnapshotID         string
	TraceID            string
	ActionID           string
	Phase              string
	Operation          string
	Target             string
	Milestone          string
	EventType          string
	Payload            string
	CreatedAt          int64
}

type SideEffectReplayPolicy string

const (
	SideEffectReplayPolicyReadOnly        SideEffectReplayPolicy = "read_only"
	SideEffectReplayPolicyIdempotentWrite SideEffectReplayPolicy = "idempotent_write"
	SideEffectReplayPolicyNonReplayable   SideEffectReplayPolicy = "non_replayable"
)

func (p SideEffectReplayPolicy) Valid() bool {
	switch p {
	case SideEffectReplayPolicyReadOnly,
		SideEffectReplayPolicyIdempotentWrite,
		SideEffectReplayPolicyNonReplayable:
		return true
	default:
		return false
	}
}

func (p SideEffectReplayPolicy) AllowsAutomaticRecovery() bool {
	return p == SideEffectReplayPolicyReadOnly ||
		p == SideEffectReplayPolicyIdempotentWrite
}

type SideEffectLedgerStatus string

const (
	SideEffectLedgerStatusPrepared    SideEffectLedgerStatus = "prepared"
	SideEffectLedgerStatusExecuting   SideEffectLedgerStatus = "executing"
	SideEffectLedgerStatusSucceeded   SideEffectLedgerStatus = "succeeded"
	SideEffectLedgerStatusFailed      SideEffectLedgerStatus = "failed"
	SideEffectLedgerStatusUnknown     SideEffectLedgerStatus = "unknown"
	SideEffectLedgerStatusCompensated SideEffectLedgerStatus = "compensated"
)

func (s SideEffectLedgerStatus) Valid() bool {
	switch s {
	case SideEffectLedgerStatusPrepared,
		SideEffectLedgerStatusExecuting,
		SideEffectLedgerStatusSucceeded,
		SideEffectLedgerStatusFailed,
		SideEffectLedgerStatusUnknown,
		SideEffectLedgerStatusCompensated:
		return true
	default:
		return false
	}
}

func (s SideEffectLedgerStatus) IsTerminal() bool {
	switch s {
	case SideEffectLedgerStatusSucceeded,
		SideEffectLedgerStatusFailed,
		SideEffectLedgerStatusUnknown,
		SideEffectLedgerStatusCompensated:
		return true
	default:
		return false
	}
}

type SideEffectResolutionAction string

const (
	SideEffectResolutionActionMarkSucceeded SideEffectResolutionAction = "mark_succeeded"
	SideEffectResolutionActionSkip          SideEffectResolutionAction = "skip"
	SideEffectResolutionActionRetry         SideEffectResolutionAction = "retry"
)

func (a SideEffectResolutionAction) Valid() bool {
	switch a {
	case SideEffectResolutionActionMarkSucceeded,
		SideEffectResolutionActionSkip,
		SideEffectResolutionActionRetry:
		return true
	default:
		return false
	}
}

type SideEffectLedger struct {
	ID                       int64
	ThreadID                 int64
	JournalRunID             int64
	AttemptID                string
	IdempotencyKey           string
	ActionKind               string
	ReplayPolicy             SideEffectReplayPolicy
	Status                   SideEffectLedgerStatus
	RequestHash              string
	RequestSummary           string
	ExternalReferenceDigest  string
	ResultSnapshotID         string
	ResultEventID            *int64
	CheckpointID             *int64
	CompensationKind         string
	ResolutionAction         SideEffectResolutionAction
	ResolutionIdempotencyKey string
	Version                  uint64
	PreparedAt               int64
	ExecutingAt              *int64
	SucceededAt              *int64
	FailedAt                 *int64
	UnknownAt                *int64
	CompensatedAt            *int64
	ResolvedAt               *int64
	CreatedAt                int64
	UpdatedAt                int64
}

type JournalSnapshotContentType string

const (
	JournalSnapshotContentTypeDocument JournalSnapshotContentType = "document"
	JournalSnapshotContentTypeTerminal JournalSnapshotContentType = "terminal"
	JournalSnapshotContentTypeCode     JournalSnapshotContentType = "code"
	JournalSnapshotContentTypeSkill    JournalSnapshotContentType = "skill"
	JournalSnapshotContentTypeBrowser  JournalSnapshotContentType = "browser"
)

func (t JournalSnapshotContentType) Valid() bool {
	switch t {
	case JournalSnapshotContentTypeDocument, JournalSnapshotContentTypeTerminal,
		JournalSnapshotContentTypeCode, JournalSnapshotContentTypeSkill,
		JournalSnapshotContentTypeBrowser:
		return true
	default:
		return false
	}
}

type JournalContentStatus string

const (
	JournalContentStatusEmpty        JournalContentStatus = "empty"
	JournalContentStatusLoading      JournalContentStatus = "loading"
	JournalContentStatusStreaming    JournalContentStatus = "streaming"
	JournalContentStatusReady        JournalContentStatus = "ready"
	JournalContentStatusError        JournalContentStatus = "error"
	JournalContentStatusNoPermission JournalContentStatus = "no_permission"
)

func (s JournalContentStatus) Valid() bool {
	switch s {
	case JournalContentStatusEmpty, JournalContentStatusLoading,
		JournalContentStatusStreaming, JournalContentStatusReady,
		JournalContentStatusError, JournalContentStatusNoPermission:
		return true
	default:
		return false
	}
}

type JournalSnapshotCompression string

const (
	JournalSnapshotCompressionIdentity JournalSnapshotCompression = "identity"
)

func (c JournalSnapshotCompression) Valid() bool {
	return c == JournalSnapshotCompressionIdentity
}

type JournalSnapshotCleanupState string

const (
	JournalSnapshotCleanupStateActive   JournalSnapshotCleanupState = "active"
	JournalSnapshotCleanupStatePending  JournalSnapshotCleanupState = "pending"
	JournalSnapshotCleanupStateDeleting JournalSnapshotCleanupState = "deleting"
	JournalSnapshotCleanupStateFailed   JournalSnapshotCleanupState = "failed"
)

func (s JournalSnapshotCleanupState) Valid() bool {
	switch s {
	case JournalSnapshotCleanupStateActive, JournalSnapshotCleanupStatePending,
		JournalSnapshotCleanupStateDeleting, JournalSnapshotCleanupStateFailed:
		return true
	default:
		return false
	}
}

type JournalContentSnapshot struct {
	SnapshotID         string
	SpaceID            int64
	ThreadID           int64
	RunID              int64
	JournalRunID       int64
	AttemptID          string
	EventID            int64
	ActionID           string
	Revision           uint32
	ContentType        JournalSnapshotContentType
	Status             JournalContentStatus
	IsFragmented       bool
	FragmentCount      uint32
	Visibility         JournalVisibility
	ErrorCode          string
	MIMEType           string
	Encoding           string
	Compression        JournalSnapshotCompression
	ContentJSON        string
	ObjectKey          string
	SummaryJSON        string
	SummaryHash        string
	ContentLength      int64
	ContentHash        string
	ACLDomain          string
	SourceResourceType string
	SourceResourceID   string
	SourceRevision     string
	OriginalObjectKey  string
	ExpiresAt          int64
	CleanupState       JournalSnapshotCleanupState
	DeletedAt          int64
	CreatedAt          int64
	Fragments          []*JournalSnapshotFragment
}

const JournalSnapshotRetentionMillis int64 = 30 * 24 * 60 * 60 * 1000

type JournalSnapshotReservation struct {
	SnapshotID       string
	ReservationToken string
	SpaceID          int64
	ThreadID         int64
	RunID            int64
	JournalRunID     int64
	AttemptID        string
	ActionID         string
	Revision         uint32
	EventID          int64
	IdempotencyKey   string
	ContentHash      string
	ACLDomain        string
	StagingPrefix    string
	ExpiresAt        int64
	CreatedAt        int64
}

type JournalSnapshotFragment struct {
	FragmentID    string
	SnapshotID    string
	FragmentIndex int32
	Kind          JournalSnapshotFragmentKind
	MetadataJSON  string
	MIMEType      string
	InlineContent string
	ObjectKey     string
	ByteStart     int64
	ByteEnd       int64
	SizeBytes     int64
	ContentHash   string
	CreatedAt     int64
}

type JournalSnapshotFragmentKind string

const (
	JournalSnapshotFragmentKindDocumentBlock    JournalSnapshotFragmentKind = "document_block"
	JournalSnapshotFragmentKindDocumentChapters JournalSnapshotFragmentKind = "document_chapters"
	JournalSnapshotFragmentKindTerminalStdout   JournalSnapshotFragmentKind = "terminal_stdout"
	JournalSnapshotFragmentKindTerminalStderr   JournalSnapshotFragmentKind = "terminal_stderr"
	JournalSnapshotFragmentKindCodeLines        JournalSnapshotFragmentKind = "code_lines"
	JournalSnapshotFragmentKindCodeHighlights   JournalSnapshotFragmentKind = "code_highlights"
	JournalSnapshotFragmentKindSkillItems       JournalSnapshotFragmentKind = "skill_items"
	JournalSnapshotFragmentKindBrowserThumbnail JournalSnapshotFragmentKind = "browser_thumbnail"
	JournalSnapshotFragmentKindBrowserSnapshot  JournalSnapshotFragmentKind = "browser_snapshot"
	JournalSnapshotFragmentKindBrowserAnalysis  JournalSnapshotFragmentKind = "browser_analysis"
)

func (k JournalSnapshotFragmentKind) Valid() bool {
	switch k {
	case JournalSnapshotFragmentKindDocumentBlock,
		JournalSnapshotFragmentKindDocumentChapters,
		JournalSnapshotFragmentKindTerminalStdout,
		JournalSnapshotFragmentKindTerminalStderr,
		JournalSnapshotFragmentKindCodeLines,
		JournalSnapshotFragmentKindCodeHighlights,
		JournalSnapshotFragmentKindSkillItems,
		JournalSnapshotFragmentKindBrowserThumbnail,
		JournalSnapshotFragmentKindBrowserSnapshot,
		JournalSnapshotFragmentKindBrowserAnalysis:
		return true
	default:
		return false
	}
}

type JournalSnapshotAction string

const (
	JournalSnapshotActionReadMetadata     JournalSnapshotAction = "read_metadata"
	JournalSnapshotActionReadContent      JournalSnapshotAction = "read_content"
	JournalSnapshotActionCopyCommand      JournalSnapshotAction = "copy_command"
	JournalSnapshotActionCopyOutput       JournalSnapshotAction = "copy_output"
	JournalSnapshotActionCopyCode         JournalSnapshotAction = "copy_code"
	JournalSnapshotActionOpenOriginal     JournalSnapshotAction = "open_original"
	JournalSnapshotActionDownloadFragment JournalSnapshotAction = "download_fragment"
)

func (a JournalSnapshotAction) Valid() bool {
	switch a {
	case JournalSnapshotActionReadMetadata, JournalSnapshotActionReadContent,
		JournalSnapshotActionCopyCommand, JournalSnapshotActionCopyOutput,
		JournalSnapshotActionCopyCode, JournalSnapshotActionOpenOriginal,
		JournalSnapshotActionDownloadFragment:
		return true
	default:
		return false
	}
}

func (a JournalSnapshotAction) UserAction() bool {
	switch a {
	case JournalSnapshotActionCopyCommand, JournalSnapshotActionCopyOutput,
		JournalSnapshotActionCopyCode, JournalSnapshotActionOpenOriginal,
		JournalSnapshotActionDownloadFragment:
		return true
	default:
		return false
	}
}

type JournalSnapshotPermissionResult string

const (
	JournalSnapshotPermissionAllowed JournalSnapshotPermissionResult = "allowed"
	JournalSnapshotPermissionDenied  JournalSnapshotPermissionResult = "denied"
	JournalSnapshotPermissionExpired JournalSnapshotPermissionResult = "expired"
)

type JournalSnapshotAccessAudit struct {
	ID               int64
	SpaceID          int64
	ThreadID         int64
	RunID            int64
	AttemptID        string
	SnapshotID       string
	ContentType      JournalSnapshotContentType
	Action           JournalSnapshotAction
	ActorID          int64
	PermissionResult JournalSnapshotPermissionResult
	IdempotencyKey   string
	TargetHash       string
	TraceID          string
	CreatedAt        int64
}

File: frontend/packages/arch/api-schema/src/idl/workbench/journal.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as base from './../base';
export { base };
export const JOURNAL_SCHEMA_VERSION = "1.1";
export const JOURNAL_PAYLOAD_VERSION = "1.0";
export const JOURNAL_PROTOCOL_VERSION = "1.1";
export const JOURNAL_SPLIT_RATIO_MIN = 0.4;
export const JOURNAL_SPLIT_RATIO_MAX = 0.7;
export enum JournalExecutionStatus {
  Pending = "pending",
  Running = "running",
  Completed = "completed",
  Failed = "failed",
  Cancelled = "cancelled",
  TimedOut = "timed_out",
  Interrupted = "interrupted",
}
export enum JournalContentStatus {
  Empty = "empty",
  Loading = "loading",
  Streaming = "streaming",
  Ready = "ready",
  Error = "error",
  NoPermission = "no_permission",
}
export enum JournalProjectionState {
  Healthy = "healthy",
  Degraded = "degraded",
  Disabled = "disabled",
}
export enum JournalVisibility {
  User = "user",
}
export enum JournalMilestonePayloadType {
  Milestone = "milestone",
}
export enum JournalActionPayloadType {
  Generic = "generic",
  Document = "document",
  Terminal = "terminal",
  Code = "code",
  Skill = "skill",
  Browser = "browser",
}
export enum JournalArtifactPayloadType {
  Artifact = "artifact",
}
export enum JournalVerificationPayloadType {
  Verification = "verification",
}
export enum JournalConfirmationPayloadType {
  Confirmation = "confirmation",
}
export enum JournalSnapshotContentType {
  Document = "document",
  Terminal = "terminal",
  Code = "code",
  Skill = "skill",
  Browser = "browser",
}
export enum JournalSnapshotFragmentKind {
  DocumentBlock = "document_block",
  DocumentChapters = "document_chapters",
  TerminalStdout = "terminal_stdout",
  TerminalStderr = "terminal_stderr",
  CodeLines = "code_lines",
  CodeHighlights = "code_highlights",
  SkillItems = "skill_items",
  BrowserThumbnail = "browser_thumbnail",
  BrowserSnapshot = "browser_snapshot",
  BrowserAnalysis = "browser_analysis",
}
export enum JournalSnapshotAction {
  CopyCommand = "copy_command",
  CopyOutput = "copy_output",
  CopyCode = "copy_code",
  OpenOriginal = "open_original",
  DownloadFragment = "download_fragment",
}
export enum JournalErrorCode {
  JournalCursorExpired = "JOURNAL_CURSOR_EXPIRED",
  JournalEventGap = "JOURNAL_EVENT_GAP",
  SnapshotUnavailable = "SNAPSHOT_UNAVAILABLE",
  ResourceNotFound = "RESOURCE_NOT_FOUND",
  RecoveryConflict = "RECOVERY_CONFLICT",
  RecoveryConfirmRequired = "RECOVERY_CONFIRM_REQUIRED",
  JournalRateLimited = "JOURNAL_RATE_LIMITED",
  SchemaIncompatible = "SCHEMA_INCOMPATIBLE",
  NoPermission = "NO_PERMISSION",
}
export enum JournalControlFrameType {
  JournalDisabled = "journal_disabled",
  JournalDegraded = "journal_degraded",
  CapabilityUnavailable = "capability_unavailable",
  ProtocolIncompatible = "protocol_incompatible",
}
export enum JournalStreamFrameKind {
  Event = "event",
  Heartbeat = "heartbeat",
  Control = "control",
}
export interface JournalEvent {
  event_id: string,
  thread_id: string,
  run_id: string,
  event_type: string,
  payload: any,
  created_at: string,
  schema_version?: string,
  attempt_id?: string,
  sequence?: number,
  idempotency_key?: string,
  parent_event_id?: string,
  status?: JournalExecutionStatus,
  occurred_at?: string,
  visibility?: JournalVisibility,
  payload_version?: string,
  snapshot_id?: string,
  trace_id?: string,
}
export interface JournalMilestoneEventData {
  milestone_id: string,
  title: string,
}
export interface JournalActionEventData {
  action_id: string,
  milestone_id?: string,
  operation: string,
  target: string,
  display_verb_running: string,
  display_verb_completed: string,
  content_type?: JournalSnapshotContentType,
}
export interface JournalArtifactEventData {
  artifact_id: string,
  collection_id?: string,
}
export interface JournalVerificationEventData {
  verification_id: string,
  title: string,
  result_summary: string,
}
export interface JournalConfirmationEventData {
  confirmation_id: string,
  confirmation_type: string,
  prompt: string,
  allowed_action_keys: string[],
}
export interface JournalMilestoneEventPayload {
  type: JournalMilestonePayloadType,
  data: JournalMilestoneEventData,
}
export interface JournalActionEventPayload {
  type: JournalActionPayloadType,
  data: JournalActionEventData,
}
export interface JournalArtifactEventPayload {
  type: JournalArtifactPayloadType,
  data: JournalArtifactEventData,
}
export interface JournalVerificationEventPayload {
  type: JournalVerificationPayloadType,
  data: JournalVerificationEventData,
}
export interface JournalConfirmationEventPayload {
  type: JournalConfirmationPayloadType,
  data: JournalConfirmationEventData,
}
export interface JournalSkill {
  skill_id: string,
  name: string,
  invocation_status?: string,
  input_summary?: string,
  output_artifacts?: string[],
  purpose_summary?: string,
  description?: string,
}
export interface JournalSnapshotFragment {
  fragment_id: string,
  fragment_index: number,
  content?: string,
  byte_start?: number,
  byte_end?: number,
  size_bytes?: number,
  content_hash?: string,
  kind?: JournalSnapshotFragmentKind,
  block_id?: string,
  stream?: string,
  start_line?: number,
  end_line?: number,
  item_start?: number,
  item_end?: number,
  binary_content_base64?: string,
  mime_type?: string,
  chapters?: JournalDocumentChapter[],
  highlights?: JournalCodeHighlight[],
  skills?: JournalSkill[],
  analysis?: string[],
}
export interface JournalDocumentChapter {
  chapter_id: string,
  title: string,
  level: number,
}
export interface JournalDocumentSnapshotContent {
  title: string,
  format?: string,
  content?: string,
  source_artifact_id?: string,
  token?: string,
  chapters?: JournalDocumentChapter[],
  active_block?: string,
  revision?: string,
  sync_status?: string,
}
export interface JournalTerminalSnapshotContent {
  command: string,
  output?: string,
  exit_code?: number,
  working_directory?: string,
  session_id?: string,
  started_at?: string,
  finished_at?: string,
  stdout?: string,
  stderr?: string,
  duration_ms?: number,
}
export interface JournalCodeHighlight {
  start_line: number,
  end_line: number,
  kind?: string,
}
export interface JournalCodeSnapshotContent {
  file_path: string,
  language?: string,
  content?: string,
  diff?: string,
  start_line?: number,
  end_line?: number,
  repository: string,
  revision: string,
  highlights?: JournalCodeHighlight[],
}
export interface JournalSkillSnapshotContent {
  skills: JournalSkill[]
}
export interface JournalBrowserSnapshotContent {
  url?: string,
  title?: string,
  screenshot_artifact_id?: string,
  capture_id: string,
  thumbnail_base64?: string,
  static_snapshot_base64: string,
  mime_type: string,
  analysis?: string[],
  index?: number,
  total?: number,
  redacted: boolean,
  redaction_evidence_id: string,
  redaction_policy_version: string,
}
export interface JournalSnapshotContent {
  document?: JournalDocumentSnapshotContent,
  terminal?: JournalTerminalSnapshotContent,
  code?: JournalCodeSnapshotContent,
  skill?: JournalSkillSnapshotContent,
  browser?: JournalBrowserSnapshotContent,
}
export interface JournalSnapshotEnvelope {
  content_type: JournalSnapshotContentType,
  snapshot_id: string,
  event_id: string,
  attempt_id: string,
  is_fragmented: boolean,
  status: JournalContentStatus,
  created_at: string,
  visibility: JournalVisibility,
  error_code?: string,
  fragments: JournalSnapshotFragment[],
  has_more: boolean,
  next_cursor?: string,
  content?: JournalSnapshotContent,
}
export interface JournalRecoveryCapability {
  allowed: boolean,
  requires_confirmation: boolean,
  allowed_actions: string[],
  reason_code?: string,
}
export interface JournalAttemptSummary {
  attempt_id: string,
  run_id: string,
  status: JournalExecutionStatus,
  projection_state: JournalProjectionState,
  latest_sequence: number,
  created_at: string,
  started_at?: string,
  ended_at?: string,
  recovery_capability: JournalRecoveryCapability,
}
export interface JournalEventPage {
  data: JournalEvent[],
  has_more: boolean,
  next_after_event_id?: string,
  attempt_id?: string,
  latest_sequence?: number,
  next_after_sequence?: number,
}
export interface JournalEnrollment {
  enrolled: boolean,
  schema_version: string,
  payload_version: string,
  journal_protocol_version: string,
  journal_enabled: boolean,
  snapshots_enabled: boolean,
}
export interface JournalBootstrap {
  attempts: JournalAttemptSummary[],
  default_attempt_id: string,
  default_attempt?: JournalAttemptSummary,
  projection_state: JournalProjectionState,
  latest_sequence: number,
  events: JournalEventPage,
  content_types: JournalSnapshotContentType[],
  enrollment: JournalEnrollment,
  submit_at: string,
  server_time: string,
  recovery_capability: JournalRecoveryCapability,
}
export interface JournalStreamMetadata {
  thread_id: string,
  run_id: string,
  attempt_id: string,
  latest_sequence: number,
  submit_at: string,
  server_time: string,
  journal_enabled?: boolean,
  snapshots_enabled?: boolean,
  journal_protocol_version: string,
}
export interface JournalControlFrame {
  type: JournalControlFrameType,
  schema_version: string,
  journal_protocol_version: string,
  server_time: string,
  attempt_id?: string,
  latest_sequence?: number,
  error_code?: JournalErrorCode,
  retryable?: boolean,
}
export interface JournalHeartbeatFrame {
  server_time: string,
  attempt_id: string,
  latest_sequence: number,
}
export interface JournalEventStreamFrame {
  kind: JournalStreamFrameKind,
  event: JournalEvent,
}
export interface JournalHeartbeatStreamFrame {
  kind: JournalStreamFrameKind,
  heartbeat: JournalHeartbeatFrame,
}
export interface JournalControlStreamFrame {
  kind: JournalStreamFrameKind,
  control: JournalControlFrame,
}
export interface GetCanonicalRunJournalRequest {
  thread_id: string,
  run_id: string,
  "X-Coze-Space-ID": string,
  attempt_id?: string,
  after_sequence?: number,
  limit?: number,
  journal_protocol_version?: string,
  after_event_id?: string,
}
export interface GetCanonicalRunSnapshotRequest {
  thread_id: string,
  run_id: string,
  snapshot_id: string,
  "X-Coze-Space-ID": string,
  cursor?: string,
  limit?: number,
}
export interface AuditCanonicalRunSnapshotActionRequest {
  thread_id: string,
  run_id: string,
  snapshot_id: string,
  "X-Coze-Space-ID": string,
  action: JournalSnapshotAction,
  "Idempotency-Key": string,
  fragment_id?: string,
}
export interface JournalSnapshotActionAuditResponse {
  snapshot_id: string,
  action: JournalSnapshotAction,
  allowed: boolean,
  audited_at: string,
  copy_text?: string,
  download_url?: string,
  download_content_base64?: string,
  download_mime_type?: string,
}
export interface RecoverCanonicalRunJournalRequest {
  thread_id: string,
  run_id: string,
  "X-Coze-Space-ID": string,
  source_attempt_id?: string,
  action: string,
  confirmed?: boolean,
  "Idempotency-Key": string,
}
export interface RecoverCanonicalRunJournalResponse {
  attempt: JournalAttemptSummary,
  accepted: boolean,
}
export interface JournalUserSettings {
  split_ratio: number,
  revision: string,
  updated_at?: string,
}
export interface GetCanonicalJournalSettingsRequest {}
export interface PatchCanonicalJournalSettingsRequest {
  split_ratio: number,
  revision: string,
}
export interface JournalErrorResponse {
  detail: string,
  error_code: string,
  code: string,
  trace_id: string,
  retryable: boolean,
}
